### PR TITLE
Improve Performance on local development platform

### PR DESCRIPTION
**Pull request by raumi75**
Hi,

I'm unexperienced on Github. So I'm not sure if a _pull request_ is the right kind of message to ask for a code change.

I develop on a small laptop, so bandwidth is not a problem, but cpu is scarce. When using hundreds of images that are rather large (850px width), I get very large blog-post sizes. That really slowed down the articles-overview to the point where it took minutes to display the page.

That's why I replaced the base64 inline image with the original image but only different width in the img-tag.

Also, because I use hundreds of image-files, it saves a lot of time to simply reuse the cached images. Why did you remove that feature? You wrote, that it feels safer. Maybe we can add an option for that?

### CODE REVIEW

**Comment by ymph** [comment]
Hello,

jus a another quick thought on my comments.
I stand on the point that the mode you suggest should be activated only with an option.

However, it also can be activated on each image differently.
The case when it clearly works (with the proper warnings in the documentation) your approach is used (when the resize format is ok, and when there is no conflict with the width and heights opassed as options), and when it do not work, emit a warning for the user and use the "old" mechanism, instead of havingerrors and stopping he rendering,

Regards,

ymh

File: lib/middleman-simple-thumbnailer/image.rb [hunk 1]
```diff
@@ -89,7 +94,7 @@ def cached_resized_img_abs_path
     end
 
     def cached_thumbnail_available?
-      File.exist?(cached_resized_img_abs_path)
+      File.exist?(cached_resized_img_abs_path) && (File.mtime(cached_resized_img_abs_path) >= File.mtime(abs_path))
```
Comment (raumi75) [inline lib/middleman-simple-thumbnailer/image.rb]: No I did not encounter any problem. It was also "better safe than sorry" thinking. So this might not be necessary.

File: lib/middleman-simple-thumbnailer/extension.rb [hunk 2]
```diff
@@ -37,12 +37,35 @@ def after_build(builder)
...
-        image = MiddlemanSimpleThumbnailer::Image.new(path, resize_to, app)
         if app.development?
-          super("data:#{image.mime_type};base64,#{image.base64_data}", options)
+
+          # For better performance on local development
+          # (high bandwidth)
+          # simply change width in <img>-tag
+          
+          original_width  = options.delete(:width)
+          resized_width   = resize_to.split('x').first.to_i
```
Comment (raumi75) [inline lib/middleman-simple-thumbnailer/extension.rb]: I see my blunder. A good solution can be quite complex.

File: lib/middleman-simple-thumbnailer/extension.rb [hunk 3]
```diff
@@ -37,12 +37,35 @@ def after_build(builder)
...
+          original_height = options.delete(:height)
+          resized_height  = resize_to.split('x').last.to_i
+          
+          if !original_width.nil? && original_width <= resized_width
+            if original_height <= resized_height
+              options[:width] = original_width
```
Comment (raumi75) [inline lib/middleman-simple-thumbnailer/extension.rb]: Ok. I understand that changing the width and height is not a sound solution. Could you explain why the image is base64-included in the first place? Is that because the cached file is not available to the web-client?

Comment (ymph) [inline lib/middleman-simple-thumbnailer/extension.rb]: As I tried to tell  you, I am not the original author of this extension.
However I think that the goal was to have a good preview of the result during development.
I have looked to middleman, and the "preview server" is an instance of the "standard" ruby "webrick" server. This http server is quite powerful and allow you to 'mount' servlets like file handler to path. 

Therefore at the extension initialisation, if we could grab the instance of the web server we could map a special path (/__thumbnail_cache for example) to the location of the cache folder.

This would not solve the fact that the images would need to be resized at least once, but it would solve the problem of having heavy data url on the page...

Comment (raumi75) [inline lib/middleman-simple-thumbnailer/extension.rb]: I understand. I have never written a middleman extension and I am totally new to ruby. But this sounds like the best solution. Maybe this path-topic will be a good solution and maybe we can work out a way to also solve Issue #13.

Comment (ymph) [inline lib/middleman-simple-thumbnailer/extension.rb]: To be honest, this is my first too...

This is not really related to #13, but could be important.

Just a quick question: What is your use case for this extension ?
Do you have a bunch of images in a well-known location that need to be resized with the same size, or do you have a lot of different images with different resizing need each time ?

Thank you,

ymh

Comment (raumi75) [inline lib/middleman-simple-thumbnailer/extension.rb]: My use case is rather simple. I scale all images in certain folders down to a width of 850px. I then create an `<a href="image.jpg"></a>` tag. Since I use the bootstrap class "image-responsive" for the image, I don't care much about the width- and height -tag.

So this could be accomplished with a script-shell `find | convert`-style solution. But I wanted to automate this at build-time and keep the scaled down-versions out of my git-repository.

After my crude change to the code, it behaves exactly the way I want it. (Having hundreds of megabytes included into blog-code that will be processed with `articles.each do` slowed things down and running imagemagick with every build drained my battery.)

Sidenote: I was thinking about using middleman-automatic-clown-car but that does not seem to be compatible with middleman 4 (at first glance).

**Comment by raumi75** [comment]
PR #16 totally solves my problem. Thank you!

The rest of this pull request was due to my inexperience with middleman. I solved the performance issue during development by moving "activate :middleman_simple_thumbnailer" into the :build codeblock of config.rb.

**Comment by ymph** [comment]
Thank you for your PR.

If it was not merged, it raised some very good points.
I am going to open an issue to explore adding a virtual path to the preview server during development.

ymph
